src/runtime-rs/crates/resource/src/volume/block_volume.rs

```diff
@@ -35,76 +38,97 @@ impl BlockVolume {
+                {
+                    warn!(sl!(), "invalid volume device file: {}", v.device);
+                    return Err(anyhow!("create direct volume failed {:?}", "OK"));
+                }
+
+                blk_dev_fstype = v.fs_type.clone();
+
+                BlockConfig {
+                    path_on_host: v.device,
+                    ..Default::default()
+                }
+            }
+            _ => {
+                return Err(anyhow!(
+                    "unsupport direct block volume r#type: {:?}",
+                    m.r#type.as_str()
+                ))
+            }
         };
 
-        let device_id = d
-            .write()
-            .await
-            .new_device(&DeviceConfig::BlockCfg(block_device_config.clone()))
+        // create and insert block device into Kata VM
+        let device_info = do_handle_device(d, &DeviceConfig::BlockCfg(block_device_config.clone()))
             .await
-            .context("failed to create deviec")?;
+            .context("do handle device failed.")?;
 
-        d.write()
+        // generate host guest shared path
+        let guest_path = generate_shared_path(m.destination.clone(), read_only, cid, sid)
             .await
-            .try_add_device(device_id.as_str())
-            .await
-            .context("failed to add deivce")?;
-
-        let file_name = Path::new(&m.source).file_name().unwrap().to_str().unwrap();
-        let file_name = generate_mount_path(cid, file_name);
-        let guest_path = do_get_guest_path(&file_name, cid, true, false);
-        let host_path = do_get_host_path(&file_name, sid, cid, true, read_only);
-        fs::create_dir_all(&host_path)
-            .map_err(|e| anyhow!("failed to create rootfs dir {}: {:?}", host_path, e))?;
-
-        // get complete device information
-        let dev_info = d
-            .read()
-            .await
-            .get_device_info(&device_id)
-            .await
-            .context("failed to get device info")?;
+            .context("generate host-guest shared path failed")?;
 
         // storage
-        let mut storage = Storage::default();
-
-        if let DeviceConfig::BlockCfg(config) = dev_info {
-            storage.driver = config.driver_option;
-            storage.source = config.virt_path;
-        }
+        let mut storage = agent::Storage {
+            mount_point: guest_path.clone(),
+            ..Default::default()
+        };
 
         storage.options = if read_only {
             vec!["ro".to_string()]
         } else {
             Vec::new()
         };
 
-        storage.mount_point = guest_path.clone();
+        let mut device_id: String = "".to_owned();
+        if let DeviceType::Block(device) = device_info {
+            // blk, mmioblk
+            storage.driver = device.config.driver_option;
+            // /dev/vdX
+            storage.source = device.config.virt_path;
+            device_id = device.device_id;
+        }
```

Comment:
From the context, there can be no else branch. It is recommended to state or add protection.

Reply:
add some comment and here is safe here as device info is correct when it's here.